pkg/lib/csv/replace_finder.go

```diff
@@ -0,0 +1,65 @@
+package csv
+
+import (
+	"github.com/operator-framework/operator-lifecycle-manager/pkg/api/apis/operators/v1alpha1"
+	"github.com/operator-framework/operator-lifecycle-manager/pkg/api/client/clientset/versioned"
+	"github.com/sirupsen/logrus"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// NewReplaceFinder returns an instance of ReplaceFinder
+func NewReplaceFinder(logger *logrus.Logger, client versioned.Interface) ReplaceFinder {
+	return &replace{
+		logger: logger,
+		client: client,
+	}
+}
+
+// ReplaceFinder is an interface that finds the next or previous
+// ClusterServiceVersion object in the upgrade path for a given CSV.
+type ReplaceFinder interface {
+	IsBeingReplaced(in *v1alpha1.ClusterServiceVersion, csvsInNamespace map[string]*v1alpha1.ClusterServiceVersion) (replacedBy *v1alpha1.ClusterServiceVersion)
+	IsReplacing(in *v1alpha1.ClusterServiceVersion) *v1alpha1.ClusterServiceVersion
+}
+
+type replace struct {
+	logger *logrus.Logger
+	client versioned.Interface
+}
+
+// IsBeingReplaced returns the corresponding ClusterServiceVersion object that
+// is replacing the given CSV specified.
+//
+// If the corresponding ClusterServiceVersion is not found nil is returned.
+func (r *replace) IsBeingReplaced(in *v1alpha1.ClusterServiceVersion, csvsInNamespace map[string]*v1alpha1.ClusterServiceVersion) (replacedBy *v1alpha1.ClusterServiceVersion) {
```

Comment:
if the CSVSet type were exported, could just reference it here

Reply:
You mean, instead of passing the map `csvsInNamespace map[string]*v1alpha1.ClusterServiceVersion`, call `csvSet.WithNamespace` from within the function to get the list of namespaces?  

I thought that the original intention was to pass the map as a parameter so that it does not need to know how to get it, and I wanted to keep that loose coupling.

I also was thinking of the following use case for my next PR on cluster operator
```
selector := labels.SelectorFromSet(labels.Set{
  "olm.clusteroperator.name": "packageserver",
})
related := csvSet.WithNamespaceAndLabels(in.GetNamespace(), v1alpha1.CSVPhaseAny, selector)
replacedBy := replaceFinder.IsBeingReplaced(in, related)
```
Thoughts?

Reply:
@ecordell I was going to make this change but then came across this - https://github.com/operator-framework/operator-lifecycle-manager/blob/master/pkg/controller/operators/olm/operator_test.go#L4321